### ISSUE TITLE
ChatGPTAPI 429 Error on second message sent

### ISSUE THREAD

**Issue by SettingUp56** [transitive-bullshit/chatgpt-api#202]
### Verify latest release

- [X] I verified that the issue exists in the latest `chatgpt` release

### Verify webapp is working

- [X] I verify that the ChatGPT webapp is working properly for this account.

### Environment details

- Node.js version : 19.0.0
- Ubuntu : 22.0.4.1 LTS
- Chromium 108.0.5359.124

### Describe the Bug

I'm running a simple web-app (conversationnal chat) on a VPS. The app has no problem running on my local environment (windows 11 - Chrome) but whenever I run it on my VPS (environment detailed), I have the following error : 
![image](https://user-images.githubusercontent.com/56033086/209559608-c08d461e-ab38-4f6a-ade1-a02c2d974221.png)

The first message sent (ie. the first call to the API) doesn't trigger the error and the response is correctly send back. This error ALWAYS happens on the second message send. This second message contains a **conversationID** as well as a **parentMessageId**. The browser is still open between the 2 messages, no new sessions are initialized. This errors appears instantly when the second message is sent. There is no relation between the complexity or length of the message and the error (and the first message is never sending back an error regardless of the prompt's complexity). I did try to add a delay but it didn't change anything. I'm using ChatGPTAPIBrowser with this configuration : 
``` 
const api = new ChatGPTAPIBrowser({
  email: process.env.OPENAI_EMAIL,
  password: process.env.OPENAI_PASSWORD,
  isGoogleLogin: true,
  debug: false, 
  minimize: true
}) 
```

my .env file contains email, password and the path to the browser used by Pupeteer (PUPPETEER_EXECUTABLE_PATH)

**Comment by 0xRaduan** [comment]
hey, I am having the same issue. Would be great if you can take a look, @transitive-bullshit.

**Comment by othema** [comment]
+1 same issue

**Comment by ruobingd** [comment]
+1 same issue

**Comment by pepesi** [comment]
This may help
https://help.openai.com/en/articles/5955598-is-api-usage-subject-to-any-rate-limits

**Comment by SettingUp56** [comment]
> This may help https://help.openai.com/en/articles/5955598-is-api-usage-subject-to-any-rate-limits

This is related to OpenAI API on their other models and not with chatGPT.

There is a potential fix to this issue here : https://github.com/transitive-bullshit/chatgpt-api/issues/174 but that's not really applicable if you need to keep track of a conversation :/. If anyone find another solution, letting you keeping a track of the conversation, would be nice.

**Comment by pupubird** [comment]
Same issue right  after I put in `conversationId`,  but if I just put `parentMessageId`  then it works fine

**Comment by AktharAzif** [comment]
+1 same issue

**Comment by michaelbecker-d** [comment]
+1. Encountered for the first time a few weeks ago and gotten progressively worse since - could they be experimenting with off-website access blocking by rate limiting to 0 somehow?

**Comment by transitive-bullshit** [comment]
@michaelbecker-d `demo-conversation.ts` uses the API version and will definitely run into 429s quite quickly. You should be using `demo-conversation-browser.ts` or `demo.ts` since they both use the browser-based version.

Another thing to watch out for: you should only ever have a max of one `sendMessage` request at a time per browser instance and OpenAI account.

If you try to have multiple `sendMessage` requests overlapping each other, you will get a 429 immediately.

**Comment by transitive-bullshit** [comment]
I agree that OpenAI seems to have made their rate limits stricter, but if you're using the browser-based version without overlapping `sendMessage` requests, it should work fairly consistently. This is what I'm using to power my twitter bot, which I had to shut off during holiday travel, but aside from that it's been working pretty well.

**Comment by lokwkin** [comment]
I am facing the same issue. I implemented a queue mechanism to make sure my service one send message at a time and wait for chatgpt's result before sending another, but it still get 429 on the second question I send out after each login (the first question is fine). 
In fact I even tried using the the ChatGPTBrowser to do the login part and then open the puppeteer chromium window to manually input and ask question, it still gets 429 on my second question too. However if I do exactly the same with normal chrome browser it with same IP and same login account, it works perfectly fine. 
I suspect they were somehow able to detect the type of browser, perhaps.

**Comment by SettingUp56** [comment]
I tried it with both chromium and chrome browser but still face the same issue. (429 on second message sent). However this issue only happens on my VPS and not my local environment (everything running smoothly here). I'm still trying to figure out what could be the difference between these environments that could cause the issue ...

**Comment by lokwkin** [comment]
Just managed to run it on servers with VPN. Previously I was running it on an AWS machine but fails with 429, and it doesn't help even if I switch to use another AWS machine, I guess it basically blocks all IP under AWS.

**Comment by 0xRaduan** [comment]
hey, @lokwkin – which VPN you are using? I think I have tried few of them, and none ever worked.

**Comment by transitive-bullshit** [comment]
@lokwkin Cloudflare is smart enough to recognize data center IPs, so servers from AWS / GCP / Azure / etc will likely lead to 429s very quickly unless you get lucky.

**Comment by transitive-bullshit** [comment]
If you're going to use a proxy or VPN service, try to get one that uses residential IPs.

**Comment by vhalbuquerque** [comment]
I tried this alternative using proxy that uses residential IPs. It didn't work for me, I've got 429 code error again and again

**Comment by lokwkin** [comment]
> I tried this alternative using proxy that uses residential IPs. It didn't work for me, I've got 429 code error again and again

Just as my experience, try look at whether the proxy forward the original header which might expose the origin of your server.

**Comment by oogxdd** [comment]
same here

for now solved it by calling api.refreshSession() after each call. it takes quite a lot of time but generally works

**Comment by nick4fake** [comment]
I can confirm that switching proxies helps

**Comment by oogxdd** [comment]
@nick4fake could you say what proxy are you using? I have tried setting AWS EC2 as proxy but apparently OpenAI blocks it's IP addresses

**Comment by LucasBadico** [comment]
Guys, I was with the same error, when I added the parentMessageId for second message, everything worked.

**Comment by transitive-bullshit** [comment]
If anyone else is experiencing this, calling `refreshSession` after each message resolves the issue, though it is pretty slow. Otherwise, read through the advice from me and other folks in this thread which should fix the underlying issues.

**Comment by arshtepe** [comment]
nothing helped me, even refreshSession didn't help
